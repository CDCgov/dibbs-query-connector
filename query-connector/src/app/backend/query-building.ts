"use server";

import { getDbClient } from "./dbClient";
import { NestedQuery, QueryDetailsResult } from "../queryBuilding/utils";
import { DibbsValueSet } from "../constants";
const dbClient = getDbClient();

/**
 * Getter function to grab saved query details from the DB
 * @param queryId - Query ID to grab data from the db with
 * @returns The query name, data, and conditions list from the query table
 */
export async function getSavedQueryDetails(queryId: string) {
  const id = queryId;
  const queryString = `
    select q.query_name, q.id, q.query_data, q.conditions_list
        from query q 
    where q.id = $1;
    `;

  try {
    const result = await dbClient.query(queryString, [id]);
    if (result.rows.length > 0) {
      return result.rows as unknown as QueryDetailsResult[];
    }
    console.error("No results found for query:", id);
    return [];
  } catch (error) {
    console.error("Error retrieving query", error);
  }
}

export async function saveCustomQuery(
  queryInput: NestedQuery,
  queryName: string,
  author: string,
) {
  const queryString = `
    insert into query
      values($1, $2, $3, $4)
      returning id, query_name
    `;
  const { queryData, conditionIds } = formatQueryDataForDatabase(queryInput);
  try {
    const dataToWrite = [queryName, queryData, conditionIds, author];
    const result = await dbClient.query(queryString, dataToWrite);
    if (result.rows.length > 0) {
      return result.rows as unknown as QueryDetailsResult[];
    }
    console.error("Query save failed:", dataToWrite);
    return [];
  } catch (error) {
    console.error("Error saving new query", error);
  }
}

function formatQueryDataForDatabase(frontendInput: NestedQuery) {
  const queryData: Record<string, { [valueSetId: string]: DibbsValueSet }> = {};
  const conditionIds: string[] = [];

  Object.entries(frontendInput).forEach(([conditionId, data]) => {
    queryData[conditionId] = {};
    conditionIds.push(conditionId);
    Object.values(data).forEach((vsNameToVsGroupingMap) => {
      Object.values(vsNameToVsGroupingMap).forEach((vsGrouping) => {
        const valueSetsToSave = vsGrouping.items;
        valueSetsToSave.forEach((dibbsVs) => {
          queryData[conditionId][dibbsVs.valueSetId] = dibbsVs;
        });
      });
    });
  });

  return { queryData, conditionIds };
}
