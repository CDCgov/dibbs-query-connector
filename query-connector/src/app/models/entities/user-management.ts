export enum RoleTypeValues {
  Admin = "Admin",
  SuperAdmin = "Super Admin",
  Standard = "Standard User",
}

export interface User {
  id: string;
  username: string;
  first_name: string;
  last_name: string;
  qc_role:
    | RoleTypeValues.SuperAdmin
    | RoleTypeValues.Admin
    | RoleTypeValues.Standard;
  userGroups?: UserGroup[];
  userGroupMemberships?: UserGroupMembership[];
}

export interface UserGroup {
  id: string;
  name: string;
  memberSize: number;
  querySize: number;
}

export interface UserGroupMembership {
  id: string;
  group_name: string;
  usergroup_id: string;
  is_member: boolean;
}

export interface Query {
  id: string;
  name: string;
  userGroupMemberships?: UserGroupMembership[];
}
