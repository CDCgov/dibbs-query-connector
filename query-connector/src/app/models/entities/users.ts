import { QueryTableResult } from "@/app/(pages)/queryBuilding/utils";

export enum UserRole {
  ADMIN = "Admin",
  STANDARD = "Standard",
  SUPER_ADMIN = "Super Admin",
}
export interface User {
  id: string;
  username: string;
  first_name: string;
  last_name: string;
  qc_role: UserRole;
  userGroupMemberships?: UserGroupMembership[];
}

export interface UserGroup {
  id: string;
  name: string;
  member_size: number;
  query_size: number;
  members?: User[];
  queries?: QueryTableResult[];
}

export interface UserGroupMembership {
  membership_id: string;
  usergroup_name: string;
  usergroup_id: string;
  is_member: boolean;
}

export interface Query {
  id: string;
  name: string;
  userGroupMemberships?: UserGroupMembership[];
}
