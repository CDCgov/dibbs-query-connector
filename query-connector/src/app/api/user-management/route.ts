import { NextResponse } from "next/server";
import { addUserIfNotExists } from "@/app/backend/user-management";

/**
 * API route for adding a user to the user_management table.
 * @param req - The incoming request object.
 * @returns A response indicating success or failure.
 */
export async function POST(req: Request) {
  try {
    const userToken = await req.json();

    if (!userToken?.username) {
      return NextResponse.json(
        { error: "Invalid user token" },
        { status: 400 },
      );
    }

    const user = await addUserIfNotExists(userToken);
    return NextResponse.json(user);
  } catch (error) {
    console.error("Error in user-management API:", error);
    return NextResponse.json(
      { error: "Internal Server Error" },
      { status: 500 },
    );
  }
}
