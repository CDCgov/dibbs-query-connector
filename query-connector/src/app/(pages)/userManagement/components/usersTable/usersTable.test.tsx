import { waitFor, screen, render } from "@testing-library/react";
import * as UserManagementBackend from "@/app/backend/user-management";
import { UserRole } from "@/app/models/entities/users";
import { renderWithUser, RootProviderMock } from "@/app/tests/unit/setup";
import UsersTable from "./usersTable";
import {
  mockSuperAdmin,
  mockAdmin,
  mockGroupBasic,
  mockPermissionsTab,
} from "../../test-utils";

jest.mock("next-auth/react");

jest.mock("@/app/backend/user-management", () => ({
  getUsers: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getUserGroups: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
}));

jest.mock(
  "@/app/ui/components/withAuth/WithAuth",
  () =>
    ({ children }: React.PropsWithChildren) => <div>{children}</div>,
);

describe("Super Admin view of Users Table", () => {
  const role = UserRole.SUPER_ADMIN;
  it("renders error message when no users are found", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        <UsersTable role={role} />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(screen.getByText("No users found")).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });

  it("renders table view after content is loaded", async () => {
    jest.spyOn(UserManagementBackend, "getUsers").mockResolvedValueOnce({
      items: [mockAdmin, mockSuperAdmin],
      totalItems: 2,
    });

    render(
      <RootProviderMock currentPage="/userManagement">
        <UsersTable role={role} />
      </RootProviderMock>,
    );
    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(document.body).toMatchSnapshot();

    if (mockPermissionsTab.renderContent) {
      render(mockPermissionsTab.renderContent());
    }

    expect(document.body).toHaveTextContent("Potter, Harry");
    expect(document.body).toHaveTextContent("Order of the Phoenix");
    expect(document.body).toMatchSnapshot();

    jest.restoreAllMocks();
  });

  it("renders content on tab click", async () => {
    jest.spyOn(UserManagementBackend, "getUsers").mockResolvedValueOnce({
      items: [mockAdmin, mockSuperAdmin],
      totalItems: 2,
    });
    jest.spyOn(UserManagementBackend, "getUserGroups").mockResolvedValueOnce({
      items: [mockGroupBasic],
      totalItems: 2,
    });
    const { user } = renderWithUser(
      <RootProviderMock currentPage="/userManagement">
        <UsersTable role={role} />
      </RootProviderMock>,
    );
    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(document.body).toMatchSnapshot();
    const usersTab = document.getElementsByClassName("tab__active")[0];
    expect(usersTab).toHaveTextContent("Users");
    expect(document.body).toHaveTextContent("Permissions");

    const groupsTab = document.getElementsByClassName("tab")[0];
    expect(groupsTab).toHaveTextContent("User groups");

    await user.click(groupsTab);
    expect(groupsTab).toHaveClass("tab__active");

    expect(document.body).toHaveTextContent("Assigned queries");
    expect(document.body).toMatchSnapshot();

    jest.restoreAllMocks();
  });
});

describe("Admin view of Users Table", () => {
  const role = UserRole.ADMIN;
  it("renders error message when no user groups are found", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        <UsersTable role={role} />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(screen.getByText("No user groups found")).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
    jest.restoreAllMocks();
  });

  it("fetches usergroups on page load", async () => {
    const getUserGroupsSpy = jest
      .spyOn(UserManagementBackend, "getUserGroups")
      .mockResolvedValueOnce({
        items: [mockGroupBasic],
        totalItems: 1,
      });

    render(
      <RootProviderMock currentPage="/userManagement">
        <UsersTable role={role} />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(getUserGroupsSpy).toHaveBeenCalled();
    await screen.findByText(mockGroupBasic.name);
    expect(document.body).toMatchSnapshot();
    jest.restoreAllMocks();
  });
});
