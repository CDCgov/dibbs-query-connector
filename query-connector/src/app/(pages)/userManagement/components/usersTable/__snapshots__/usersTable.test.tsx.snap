// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin view of Users Table fetches usergroups on page load 1`] = `
<body>
  <div>
    <div
      class="customizeTableWrapper"
    >
      <table
        class="usa-table usa-table--borderless customizeTable"
        data-testid="table"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Members
            </th>
            <th>
              Assigned queries
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              width="270"
            >
              Hogwarts
            </td>
            <td>
              <button
                aria-description="Edit Hogwarts members"
                class="usa-button usa-button--unstyled drawerButton text-no-underline"
                data-testid="edit-member-list-0"
                type="button"
              >
                1 member
              </button>
            </td>
            <td>
              <button
                aria-description="Edit Hogwarts queries"
                class="usa-button usa-button--unstyled text-no-underline"
                data-testid="edit-query-list-0"
                type="button"
              >
                0 queries
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="drawer closed"
        data-testid="drawer-open-false"
        role="dialog"
      >
        <div
          class="drawerContent display-none"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-2"
              data-testid="drawer-title"
            />
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="padding-top-2"
          >
            <div>
              Error: could not retrieve queries for user group. Please try again later.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Admin view of Users Table renders error message when no user groups are found 1`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="empty-response"
      >
        No user groups found
      </div>
      <div
        class="drawer closed"
        data-testid="drawer-open-false"
        role="dialog"
      >
        <div
          class="drawerContent display-none"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-2"
              data-testid="drawer-title"
            />
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="padding-top-2"
          >
            <div>
              Error: could not retrieve queries for user group. Please try again later.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders content on tab click 1`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <a
          class="tab__active"
          data-prevent-nprogress="true"
          href=""
        >
          Users
        </a>
        <a
          class="tab"
          data-prevent-nprogress="true"
          href=""
        >
          User groups
        </a>
      </div>
      <div
        class="customizeTableWrapper"
      >
        <table
          class="usa-table usa-table--borderless customizeTable"
          data-testid="table"
        >
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                Permissions
              </th>
              <th>
                User groups
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                width="270"
              >
                Potter, Lily 
          
              </td>
              <td
                class="roleDropdown"
                width="270"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-456"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-456"
                  id="role-select-456"
                >
                  <option
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    selected=""
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                <button
                  aria-description="Edit Order of the Phoenix members"
                  class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  Order of the Phoenix
                </button>
              </td>
            </tr>
            <tr>
              <td
                width="270"
              >
                Potter, Harry 
          
              </td>
              <td
                class="roleDropdown"
                width="270"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-123"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-123"
                  id="role-select-123"
                >
                  <option
                    selected=""
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                --
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="drawer closed"
        data-testid="drawer-open-false"
        role="dialog"
      >
        <div
          class="drawerContent display-none"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-2"
              data-testid="drawer-title"
            />
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="padding-top-2"
          >
            <div>
              Error: could not retrieve queries for user group. Please try again later.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders content on tab click 2`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <a
          class="tab"
          data-prevent-nprogress="true"
          href=""
        >
          Users
        </a>
        <a
          class="tab__active"
          data-prevent-nprogress="true"
          href=""
        >
          User groups
        </a>
      </div>
      <div
        class="customizeTableWrapper"
      >
        <table
          class="usa-table usa-table--borderless customizeTable"
          data-testid="table"
        >
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                Members
              </th>
              <th>
                Assigned queries
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                width="270"
              >
                Hogwarts
              </td>
              <td>
                <button
                  aria-description="Edit Hogwarts members"
                  class="usa-button usa-button--unstyled drawerButton text-no-underline"
                  data-testid="edit-member-list-0"
                  type="button"
                >
                  1 member
                </button>
              </td>
              <td>
                <button
                  aria-description="Edit Hogwarts queries"
                  class="usa-button usa-button--unstyled text-no-underline"
                  data-testid="edit-query-list-0"
                  type="button"
                >
                  0 queries
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="drawer closed"
        data-testid="drawer-open-false"
        role="dialog"
      >
        <div
          class="drawerContent display-none"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-2"
              data-testid="drawer-title"
            />
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="padding-top-2"
          >
            <div>
              Error: could not retrieve queries for user group. Please try again later.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders error message when no users are found 1`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <a
          class="tab__active"
          data-prevent-nprogress="true"
          href=""
        >
          Users
        </a>
        <a
          class="tab"
          data-prevent-nprogress="true"
          href=""
        >
          User groups
        </a>
      </div>
      <div
        class="empty-response"
      >
        No users found
      </div>
      <div
        class="drawer closed"
        data-testid="drawer-open-false"
        role="dialog"
      >
        <div
          class="drawerContent display-none"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-2"
              data-testid="drawer-title"
            />
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="padding-top-2"
          >
            <div>
              Error: could not retrieve queries for user group. Please try again later.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders table view after content is loaded 1`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <a
          class="tab__active"
          data-prevent-nprogress="true"
          href=""
        >
          Users
        </a>
        <a
          class="tab"
          data-prevent-nprogress="true"
          href=""
        >
          User groups
        </a>
      </div>
      <div
        class="customizeTableWrapper"
      >
        <table
          class="usa-table usa-table--borderless customizeTable"
          data-testid="table"
        >
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                Permissions
              </th>
              <th>
                User groups
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                width="270"
              >
                Potter, Lily 
          
              </td>
              <td
                class="roleDropdown"
                width="270"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-456"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-456"
                  id="role-select-456"
                >
                  <option
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    selected=""
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                <button
                  aria-description="Edit Order of the Phoenix members"
                  class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  Order of the Phoenix
                </button>
              </td>
            </tr>
            <tr>
              <td
                width="270"
              >
                Potter, Harry 
          
              </td>
              <td
                class="roleDropdown"
                width="270"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-123"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-123"
                  id="role-select-123"
                >
                  <option
                    selected=""
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                --
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="drawer closed"
        data-testid="drawer-open-false"
        role="dialog"
      >
        <div
          class="drawerContent display-none"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-2"
              data-testid="drawer-title"
            />
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="padding-top-2"
          >
            <div>
              Error: could not retrieve queries for user group. Please try again later.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders table view after content is loaded 2`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <a
          class="tab__active"
          data-prevent-nprogress="true"
          href=""
        >
          Users
        </a>
        <a
          class="tab"
          data-prevent-nprogress="true"
          href=""
        >
          User groups
        </a>
      </div>
      <div
        class="customizeTableWrapper"
      >
        <table
          class="usa-table usa-table--borderless customizeTable"
          data-testid="table"
        >
          <thead>
            <tr>
              <th>
                Name
              </th>
              <th>
                Permissions
              </th>
              <th>
                User groups
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td
                width="270"
              >
                Potter, Lily 
          
              </td>
              <td
                class="roleDropdown"
                width="270"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-456"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-456"
                  id="role-select-456"
                >
                  <option
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    selected=""
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                <button
                  aria-description="Edit Order of the Phoenix members"
                  class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  Order of the Phoenix
                </button>
              </td>
            </tr>
            <tr>
              <td
                width="270"
              >
                Potter, Harry 
          
              </td>
              <td
                class="roleDropdown"
                width="270"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-123"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-123"
                  id="role-select-123"
                >
                  <option
                    selected=""
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                --
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="drawer closed"
        data-testid="drawer-open-false"
        role="dialog"
      >
        <div
          class="drawerContent display-none"
        >
          <div
            class="drawerHeader"
          >
            <button
              aria-label="Close drawer"
              class="closeButton"
              data-testid="close-drawer"
            >
              <svg
                aria-label="X icon indicating closure"
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </button>
            <h2
              class="margin-0 padding-bottom-2"
              data-testid="drawer-title"
            />
            <div>
              <input
                class="usa-input searchField"
                data-testid="textInput"
                id="searchFieldTemplate"
                name="searchFieldTemplate"
                placeholder="Search"
                type="search"
                value=""
              />
            </div>
          </div>
          <div
            class="padding-top-2"
          >
            <div>
              Error: could not retrieve queries for user group. Please try again later.
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      class="customizeTableWrapper"
    >
      <table
        class="usa-table usa-table--borderless customizeTable"
        data-testid="table"
      >
        <thead>
          <tr>
            <th>
              Name
            </th>
            <th>
              Permissions
            </th>
            <th>
              User groups
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              width="270"
            >
              Potter, Lily 
          
            </td>
            <td
              class="roleDropdown"
              width="270"
            >
              <label
                class="usa-label usa-sr-only"
                data-testid="label"
                for="role-select-456"
              >
                User role
              </label>
              <select
                class="usa-select"
                data-testid="role-select-456"
                id="role-select-456"
              >
                <option
                  value="Super Admin"
                >
                  Super Admin
                </option>
                <option
                  selected=""
                  value="Admin"
                >
                  Admin
                </option>
                <option
                  value="Standard"
                >
                  Standard
                </option>
              </select>
            </td>
            <td>
              <button
                aria-description="Edit Order of the Phoenix members"
                class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                data-testid="button"
                type="button"
              >
                Order of the Phoenix
              </button>
            </td>
          </tr>
          <tr>
            <td
              width="270"
            >
              Potter, Harry 
          
            </td>
            <td
              class="roleDropdown"
              width="270"
            >
              <label
                class="usa-label usa-sr-only"
                data-testid="label"
                for="role-select-123"
              >
                User role
              </label>
              <select
                class="usa-select"
                data-testid="role-select-123"
                id="role-select-123"
              >
                <option
                  selected=""
                  value="Super Admin"
                >
                  Super Admin
                </option>
                <option
                  value="Admin"
                >
                  Admin
                </option>
                <option
                  value="Standard"
                >
                  Standard
                </option>
              </select>
            </td>
            <td>
              --
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
